watchtower/wtclient/client_test.go

```diff
@@ -2416,6 +2416,139 @@ var clientTests = []clientTest{
 			server2.waitForUpdates(hints[numUpdates/2:], waitTime)
 		},
 	},
+	{
+		// This test demonstrates a bug that will be addressed in a
+		// follow-up commit. It shows that if a channel is closed while
+		// an update for that channel still exists in an in-memory queue
+		// somewhere then it is possible that all the data for that
+		// channel gets deleted from the tower client DB. This results
+		// in an error being thrown in the DB AckUpdate method since it
+		// will try to find the associated channel data but will not
+		// find it.
+		name: "channel closed while update is un-acked",
+		cfg: harnessCfg{
+			localBalance:  localBalance,
+			remoteBalance: remoteBalance,
+			policy: wtpolicy.Policy{
+				TxPolicy:   defaultTxPolicy,
+				MaxUpdates: 5,
+			},
+		},
+		fn: func(h *testHarness) {
+			const numUpdates = 10
+
+			h.sendUpdatesOn = true
+
+			// Advance the channel with a few updates.
+			hints := h.advanceChannelN(0, numUpdates)
+
+			// Backup a few these updates and wait for them to
+			// arrive at the server. Note that we back up enough
+			// updates to saturate the session so that the session
+			// is considered closable when the channel is deleted.
+			h.backupStates(0, 0, numUpdates/2, nil)
+			h.server.waitForUpdates(hints[:numUpdates/2], waitTime)
+
+			// Now, restart the server in a state where it will not
+			// ack updates. This will allow us to wait for an
+			// update to be un-acked and persisted.
+			h.server.restart(func(cfg *wtserver.Config) {
+				cfg.NoAckUpdates = true
+			})
+
+			// Backup a few more of the update. These should remain
+			// in the client as un-acked.
+			h.backupStates(0, numUpdates/2, numUpdates-1, nil)
+
+			// Wait for the tasks to be bound to sessions.
+			fetchSessions := h.clientDB.FetchSessionCommittedUpdates
+			err := wait.Predicate(func() bool {
+				sessions, err := h.clientDB.ListClientSessions(
```

Comment:
Q: I'm assuming there should be one session and this session should have `len(updates) == numUpdates`, is that correct?

Reply:
`numUpdates-1` :) 

Cant do `numUpdates` cause then once the fix is in, the session will be seen as closable if all the committed updates are rogue which then makes it hard to do all the other assertions. 

I've changed it to check `len(updates) !=numUpdates-1` to make it more clear :)